File: tests/fixtures/fifo/before/zircon/kernel/object/fifo_dispatcher.cc
// Copyright 2016 The Fuchsia Authors
//
// Use of this source code is governed by a MIT-style
// license that can be found in the LICENSE file or at
// https://opensource.org/licenses/MIT

#include "object/fifo_dispatcher.h"

#include <string.h>
#include <zircon/rights.h>

#include <kernel/mutex.h>

zx_status_t FifoDispatcher::WriteFromUser(size_t elem_size, user_in_ptr<const uint8_t> ptr,
                                          size_t count, size_t* actual) {
  canary_.Assert();

  Guard<CriticalMutex> guard{get_lock()};
  if (!peer()) {
    return ZX_ERR_PEER_CLOSED;
  }
  return peer()->WriteSelfLocked(elem_size, ptr, count, actual);
}

zx_status_t FifoDispatcher::WriteSelfLocked(size_t elem_size, user_in_ptr<const uint8_t> ptr,
                                            size_t count, size_t* actual) {
  canary_.Assert();

  if (elem_size != elem_size_) {
    return ZX_ERR_OUT_OF_RANGE;
  }
  if (count == 0) {
    return ZX_ERR_OUT_OF_RANGE;
  }

  uint32_t old_head = head_;

  // Total number of available empty slots in the fifo.
  size_t avail = elem_count_ - (head_ - tail_);

  if (avail == 0) {
    return ZX_ERR_SHOULD_WAIT;
  }

  bool was_empty = (head_ == tail_);

  if (count > avail) {
    count = avail;
  }

  while (count > 0) {
    uint32_t offset = (head_ % elem_count_);

    // Number of slots from target to end of fifo.
    size_t n = (elem_count_ - offset);
    if (n > count) {
      n = count;
    }

    zx_status_t status = ptr.copy_array_from_user(&data_[offset * elem_size_], n * elem_size_);
    if (status != ZX_OK) {
      // Roll back, in case this is the second copy.
      head_ = old_head;
      return ZX_ERR_INVALID_ARGS;
    }

    head_ += static_cast<uint32_t>(n);
    count -= n;
    ptr = ptr.byte_offset(n * elem_size_);
  }

  if (was_empty) {
    UpdateStateLocked(0u, ZX_FIFO_READABLE);
  }

  if (IsFullLocked()) {
    UpdateStateLocked(ZX_FIFO_WRITABLE, 0u);
  }

  *actual = (head_ - old_head);
  return ZX_OK;
}

zx_status_t FifoDispatcher::ReadToUser(size_t elem_size, user_out_ptr<uint8_t> ptr, size_t count,
                                       size_t* actual) {
  canary_.Assert();

  if (elem_size != elem_size_) {
    return ZX_ERR_OUT_OF_RANGE;
  }
  if (count == 0) {
    return ZX_ERR_OUT_OF_RANGE;
  }

  Guard<CriticalMutex> guard{get_lock()};

  // Total number of available entries to read from the fifo.
  size_t avail = (head_ - tail_);

  if (avail == 0) {
    return peer() ? ZX_ERR_SHOULD_WAIT : ZX_ERR_PEER_CLOSED;
  }

  bool was_full = IsFullLocked();

  if (count > avail) {
    count = avail;
  }

  uint32_t old_tail = tail_;
  while (count > 0) {
    uint32_t offset = (tail_ % elem_count_);

    // Number of slots from target to end of fifo.
    size_t n = (elem_count_ - offset);
    if (n > count) {
      n = count;
    }

    zx_status_t status = ptr.copy_array_to_user(&data_[offset * elem_size_], n * elem_size_);
    if (status != ZX_OK) {
      // Roll back, in case this is the second copy.
      tail_ = old_tail;
      return ZX_ERR_INVALID_ARGS;
    }

    tail_ += static_cast<uint32_t>(n);
    count -= n;
    ptr = ptr.byte_offset(n * elem_size_);
  }

  // If the fifo was full, it is now writable.
  if (was_full && peer()) {
    peer()->UpdateStateLocked(0u, ZX_FIFO_WRITABLE);
  }

  *actual = (tail_ - old_tail);
  return ZX_OK;
}

bool FifoDispatcher::IsFullLocked() const { return head_ - tail_ == elem_count_; }

File: tests/fixtures/fifo/after/zircon/kernel/object/fifo_dispatcher.rs
// Copyright 2026 The Fuchsia Authors
//
// Use of this source code is governed by a MIT-style
// license that can be found in the LICENSE file or at
// https://opensource.org/licenses/MIT

use crate::object::dispatcher::{PeeredDispatcher, StateLockGuard};
use crate::user_copy::{UserInPtr, UserOutPtr};
use zx_status::Status;
use zx_types::{ZX_FIFO_READABLE, ZX_FIFO_WRITABLE};

impl FifoDispatcher {
    /// Writes `count` elements of `elem_size` bytes from `ptr` into the peer.
    ///
    /// Returns ZX_ERR_OUT_OF_RANGE if `elem_size` does not match the fifo.
    pub fn write_from_user(
        &self,
        elem_size: usize,
        ptr: UserInPtr<u8>,
        count: usize,
    ) -> Result<usize, Status> {
        self.canary.assert();

        ksync::lock!(let guard = self.lock());
        let Some(peer) = self.peer(&guard) else {
            return Err(Status::BAD_STATE);
        };
        peer.write_self_locked(&guard, elem_size, ptr, count)
    }

    fn write_self_locked(
        &self,
        guard: &StateLockGuard<'_>,
        elem_size: usize,
        mut ptr: UserInPtr<u8>,
        mut count: usize,
    ) -> Result<usize, Status> {
        self.canary.assert();

        if elem_size != self.elem_size as usize {
            return Err(Status::OUT_OF_RANGE);
        }
        if count == 0 {
            return Err(Status::OUT_OF_RANGE);
        }

        let old_head = guard.head();

        // Total number of available empty slots in the fifo.
        let avail = (self.elem_count - guard.head().wrapping_sub(guard.tail())) as usize;

        if avail == 0 {
            return Err(Status::SHOULD_WAIT);
        }

        let was_empty = guard.head() == guard.tail();

        if count > avail {
            count = avail;
        }

        while count > 0 {
            let offset = guard.head() % self.elem_count;

            // Number of slots from target to end of fifo.
            let mut n = (self.elem_count - offset) as usize;
            if n > count {
                n = count;
            }

            ptr.copy_array_from_user(self.data_mut(guard, offset, n))?;

            guard.set_head(guard.head().wrapping_add(n as u32));
            count -= n;
            ptr = ptr.byte_offset(n * self.elem_size as usize);
        }

        if was_empty {
            self.update_state_locked(guard, 0, ZX_FIFO_READABLE);
        }

        if self.is_full_locked(guard) {
            self.update_state_locked(guard, ZX_FIFO_WRITABLE, 0);
        }

        Ok(guard.head().wrapping_sub(old_head) as usize)
    }

    /// Reads up to `count` elements of `elem_size` bytes into `ptr`.
    pub fn read_to_user(
        &self,
        elem_size: usize,
        mut ptr: UserOutPtr<u8>,
        mut count: usize,
    ) -> Result<usize, Status> {
        self.canary.assert();

        ksync::lock!(let guard = self.lock());

        if elem_size != self.elem_size as usize {
            return Err(Status::OUT_OF_RANGE);
        }
        if count == 0 {
            return Err(Status::OUT_OF_RANGE);
        }

        // Total number of available entries to read from the fifo.
        let avail = guard.head().wrapping_sub(guard.tail()) as usize;

        if avail == 0 {
            return Err(if self.peer(&guard).is_some() {
                Status::SHOULD_WAIT
            } else {
                Status::PEER_CLOSED
            });
        }

        let was_full = self.is_full_locked(&guard);

        if count > avail {
            count = avail;
        }

        let old_tail = guard.tail();
        while count > 0 {
            let offset = guard.tail() % self.elem_count;

            // Number of slots from target to end of fifo.
            let mut n = (self.elem_count - offset) as usize;
            if n > count {
                n = count;
            }

            if ptr.copy_array_to_user(self.data(&guard, offset, n)).is_err() {
                // Roll back, in case this is the second copy.
                guard.set_tail(old_tail);
                return Err(Status::INVALID_ARGS);
            }

            guard.set_tail(guard.tail().wrapping_add(n as u32));
            count -= n;
            ptr = ptr.byte_offset(n * self.elem_size as usize);
        }

        // If the fifo was full, it is now writable.
        if was_full {
            if let Some(peer) = self.peer(&guard) {
                peer.update_state_locked(&guard, 0, ZX_FIFO_WRITABLE);
            }
        }

        Ok(guard.tail().wrapping_sub(old_tail) as usize)
    }

    /// Returns true if the fifo has no room for another element.
    fn is_full_locked(&self, guard: &StateLockGuard<'_>) -> bool {
        guard.head().wrapping_sub(guard.tail()) == self.elem_count
    }
}
